src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,82 +700,199 @@ namespace nbl::ext::imgui
 
 				for (int n = 0; n < drawData->CmdListsCount; n++)
 				{
-					const ImDrawList* cmd = drawData->CmdLists[n];
-					::memcpy(vertex_ptr, cmd->VtxBuffer.Data, cmd->VtxBuffer.Size * sizeof(ImDrawVert));
-					::memcpy(index_ptr, cmd->IdxBuffer.Data, cmd->IdxBuffer.Size * sizeof(ImDrawIdx));
-					vertex_ptr += cmd->VtxBuffer.Size;
-					index_ptr += cmd->IdxBuffer.Size;
-				}
+					const auto* cmd_list = drawData->CmdLists[n];
+					for (int cmd_i = 0; cmd_i < cmd_list->CmdBuffer.Size; cmd_i++)
+					{
+						const auto* pcmd = &cmd_list->CmdBuffer[cmd_i];
+
+						const auto clipRectangle = clip.getClipRectangle(pcmd);
+
+						// update mdi's indirect & element structures
+						auto* indirect = indirectsMappedPointer + drawID;
+						auto* element = elementsMappedPointer + drawID;
+
+						indirect->firstIndex = pcmd->IdxOffset + globalIOffset;
+						indirect->firstInstance = drawID; // use base instance as draw ID
+						indirect->indexCount = pcmd->ElemCount;
+						indirect->instanceCount = 1u;
+						indirect->vertexOffset = pcmd->VtxOffset + globalVOffset;
+
+						const auto scissor = clip.getScissor(clipRectangle);
+
+						auto packSnorm16 = [](float ndc) -> int16_t
+						{
+							return std::round<int16_t>(std::clamp(ndc, -1.0f, 1.0f) * 32767.0f); // TODO: ok encodePixels<asset::EF_R16_SNORM, double>(void* _pix, const double* _input) but iirc we have issues with our encode/decode utils
+						};
+
+						const auto vMin = trs.toNDC(core::vector2df_SIMD(scissor.offset.x, scissor.offset.y));
+						const auto vMax = trs.toNDC(core::vector2df_SIMD(scissor.offset.x + scissor.extent.width, scissor.offset.y + scissor.extent.height));
 
-				vBinding.memory->unmap();
-				iBinding.memory->unmap();
+						element->aabbMin.x = packSnorm16(vMin.x);
+						element->aabbMin.y = packSnorm16(vMin.y);
+						element->aabbMax.x = packSnorm16(vMax.x);
+						element->aabbMax.y = packSnorm16(vMax.y);
+						element->texId = pcmd->TextureId;
+
+						++drawID;
+					}
+
+					// update mdi's vertex & index buffers
+					::memcpy(indicesMappedPointer + globalIOffset, cmd_list->IdxBuffer.Data, cmd_list->IdxBuffer.Size * sizeof(ImDrawIdx));
+					::memcpy(verticesMappedPointer + globalVOffset, cmd_list->VtxBuffer.Data, cmd_list->VtxBuffer.Size * sizeof(ImDrawVert));
+
+					globalIOffset += cmd_list->IdxBuffer.Size;
+					globalVOffset += cmd_list->VtxBuffer.Size;
+				}
 			}
 
+			auto* rawPipeline = pipeline.get();
+			commandBuffer->bindGraphicsPipeline(rawPipeline);
+			commandBuffer->bindDescriptorSets(EPBP_GRAPHICS, rawPipeline->getLayout(), 0, 1, &descriptorSet);
 			{
 				const asset::SBufferBinding<const video::IGPUBuffer> binding =
 				{
-					.offset = 0,
-					.buffer = core::smart_refctd_ptr(indexBuffer)
+					.offset = multiAllocParams.offsets[MDI::EBC_INDEX_BUFFERS],
+					.buffer = smart_refctd_ptr(mdiBuffer)
 				};
 
 				if (!commandBuffer->bindIndexBuffer(binding, sizeof(ImDrawIdx) == 2 ? EIT_16BIT : EIT_32BIT))
```

Comment:
if you do an `overflowSubmit` you need to re-record these, so:
1. put these as far up as possible in your `render` function just after when you know youll have geometry to draw
2. but also put them into the failure case just after you call `overflowSubmit` and finish waiting on the semaphore

P.S. also the `bindVertexBuffers` , `setViewport` and `setScissor`

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/707#discussion_r1763267517